File: fsos-frontend/src/domain/access.js
// Unified access model — per ROLE, per AREA, one of: none / view / edit.
// Ported from the Frontseat snoboard `accessModel.ts`, re-keyed to FSOS roles.
// Route gating, nav visibility and edit-gating all derive from resolveAccess().
// Defaults below are a starting point — admins tune them in Users & Roles
// (role overrides + per-person overrides live in db.access and merge over these).
import { ROLES } from "./constants";

export const LEVELS = ["none", "view", "edit"];
export const LEVEL_LABEL = { none: "—", view: "View", edit: "Edit" };
const RANK = { none: 0, view: 1, edit: 2 };

export const PINTU_URL = "http://16.112.125.207:5173/";

export const AREA_GROUP_ORDER = ["Workspace", "Content tools", "Cops", "Growth", "Admin"];

// Every gate-able surface, in sidebar order. `route` is the primary path it guards.
export const AREAS = [
  { key: "command_room", label: "Command Room", group: "Workspace", route: "/" },
  { key: "bo_studio", label: "BO Studio", group: "Workspace", route: "/bo" },
  { key: "hpn_desk", label: "HPN Desk", group: "Workspace", route: "/hpn" },
  { key: "production", label: "Production", group: "Workspace", route: "/production" },
  { key: "distribution", label: "Distribution", group: "Workspace", route: "/distribution" },
  { key: "performance", label: "Performance", group: "Workspace", route: "/performance" },

  { key: "news", label: "News Feed", group: "Content tools", route: "/news" },
  { key: "pintu", label: "Pintu", group: "Content tools", route: PINTU_URL, external: true },

  { key: "six_day", label: "6-Day Tracker", group: "Cops", route: "/six-day-tracker" },

  { key: "growth", label: "Growth", group: "Growth", route: "/growth" },

  { key: "users_roles", label: "Users & Roles", group: "Admin", route: "/users-roles" },
  { key: "settings", label: "Settings", group: "Admin", route: "/settings" },
];

export const AREA_KEYS = AREAS.map((a) => a.key);

const all = (level) => Object.fromEntries(AREA_KEYS.map((k) => [k, level]));
const withOverrides = (base, overrides) => ({ ...all(base), ...overrides });

const WORKSPACE_ALL = {
  command_room: "edit", bo_studio: "edit", hpn_desk: "edit",
  production: "edit", distribution: "edit", performance: "edit",
};

// Default matrices. Workspace areas mirror the pre-existing FSOS nav per role, so
// nothing anyone could already see disappears; the new tools get conservative defaults.
export const ROLE_ACCESS_DEFAULTS = {
  "Founder/Admin": all("edit"),

  COA: all("edit"),

  "Short-form Lead": withOverrides("none", {
    ...WORKSPACE_ALL,
    news: "edit", pintu: "view", six_day: "edit", growth: "view",
  }),

  CS: withOverrides("none", {
    command_room: "edit", bo_studio: "edit", hpn_desk: "edit", production: "edit", performance: "edit",
    news: "edit", pintu: "view", six_day: "view", growth: "view",
  }),

  Designer: withOverrides("none", {
    production: "edit", pintu: "view", growth: "view",
  }),

  Editor: withOverrides("none", {
    production: "edit", pintu: "view", growth: "view",
  }),

  COC: withOverrides("none", {
    command_room: "edit", distribution: "edit", performance: "edit",
    news: "view", six_day: "edit", growth: "view",
  }),
};

// The role that can never be locked out — overrides on it are ignored for safety.
export const LOCKED_ROLE = "Founder/Admin";

/** Effective matrix for a single role (defaults + persisted role overrides). */
export function resolveRoleAccess(role, roleOverrides) {
  const base = ROLE_ACCESS_DEFAULTS[role] || all("none");
  if (role === LOCKED_ROLE) return base;
  const ov = roleOverrides?.[role];
  return ov ? { ...base, ...ov } : base;
}

/** Highest level across all roles, then the per-person matrix on top. */
export function resolvePersonAccess(roles, personAccess, roleOverrides) {
  const base = all("none");
  (roles || []).forEach((r) => {
    const m = resolveRoleAccess(r, roleOverrides);
    AREA_KEYS.forEach((k) => { if (RANK[m[k]] > RANK[base[k]]) base[k] = m[k]; });
  });
  if (roles?.includes(LOCKED_ROLE)) return base;
  if (!personAccess || !Object.keys(personAccess).length) return base;
  return { ...base, ...personAccess };
}

/**
 * Effective access for the acting user, or for whatever is being previewed.
 *
 * Preview comes in two kinds. A *role* shows that role's defaults — what a new Editor
 * would get. A *person* shows what one named teammate actually sees, which is the more
 * useful question here, because per-person overrides are exactly what Users & Roles is
 * for and a role preview is blind to them.
 */
export function resolveAccess(db, user, preview) {
  const access = db.access || {};
  if (preview?.kind === "role") return resolveRoleAccess(preview.role, access.roles);
  if (preview?.kind === "person") {
    const person = (db.users || []).find((u) => u.id === preview.id);
    if (person) return resolvePersonAccess(person.roles, access.people?.[person.id], access.roles);
  }
  return resolvePersonAccess(user?.roles, access.people?.[user?.id], access.roles);
}

/** True when the user has no usable role yet (Admin must assign one). */
export function isAwaitingAccess(user) {
  return !user?.roles?.length;
}

export function areaForRoute(pathname) {
  const exact = AREAS.find((a) => !a.external && a.route === pathname);
  if (exact) return exact.key;
  const pre = AREAS.find((a) => !a.external && a.route !== "/" && pathname.startsWith(`${a.route}/`));
  return pre?.key || null;
}

export function countLevels(matrix) {
  return {
    edit: AREA_KEYS.filter((k) => matrix[k] === "edit").length,
    view: AREA_KEYS.filter((k) => matrix[k] === "view").length,
  };
}

export const PREVIEW_ROLES = ROLES;

File: fsos-backend/app/access.py
"""FSOS access model, server side.

Mirrors fsos-frontend/src/domain/access.js — keep the two in sync. The frontend hides
what you can't use; this decides what you're actually allowed to do.
"""
from fastapi import HTTPException

AREAS = [
    "command_room", "bo_studio", "hpn_desk", "production", "distribution", "performance",
    "news", "pintu", "six_day", "growth", "users_roles", "settings",
]
LEVELS = ("none", "view", "edit")
RANK = {"none": 0, "view": 1, "edit": 2}
LOCKED_ROLE = "Founder/Admin"

_WORKSPACE = {a: "edit" for a in ("command_room", "bo_studio", "hpn_desk", "production", "distribution", "performance")}


def _m(**levels) -> dict:
    return {a: levels.get(a, "none") for a in AREAS}


ROLE_ACCESS_DEFAULTS: dict[str, dict] = {
    "Founder/Admin": {a: "edit" for a in AREAS},
    "COA": {a: "edit" for a in AREAS},
    "Short-form Lead": _m(**_WORKSPACE, news="edit", pintu="view", six_day="edit", growth="view"),
    "CS": _m(command_room="edit", bo_studio="edit", hpn_desk="edit", production="edit", performance="edit",
             news="edit", pintu="view", six_day="view", growth="view"),
    "Designer": _m(production="edit", pintu="view", growth="view"),
    "Editor": _m(production="edit", pintu="view", growth="view"),
    "COC": _m(command_room="edit", distribution="edit", performance="edit",
              news="view", six_day="edit", growth="view"),
}


def resolve_role_access(role: str, role_overrides: dict | None = None) -> dict:
    base = ROLE_ACCESS_DEFAULTS.get(role) or {a: "none" for a in AREAS}
    if role == LOCKED_ROLE:
        return base
    override = (role_overrides or {}).get(role)
    return {**base, **override} if override else base


def resolve_person_access(roles: list[str], person_override: dict | None = None, role_overrides: dict | None = None) -> dict:
    matrix = {a: "none" for a in AREAS}
    for role in roles or []:
        role_matrix = resolve_role_access(role, role_overrides)
        for area in AREAS:
            if RANK[role_matrix[area]] > RANK[matrix[area]]:
                matrix[area] = role_matrix[area]
    if LOCKED_ROLE in (roles or []):
        return matrix
    return {**matrix, **(person_override or {})} if person_override else matrix


def grants_beyond(matrix: dict, ceiling: dict, previous: dict | None = None) -> str | None:
    """The first area where `matrix` raises access above `ceiling`, otherwise None.

    Used to stop privilege escalation: you can't hand out more than you hold. Areas that
    `previous` already granted are ignored, so you can still *lower* someone who currently
    outranks you.
    """
    for area, level in matrix.items():
        rank = RANK.get(level, 0)
        if rank > RANK.get(ceiling.get(area, "none"), 0) and rank > RANK.get((previous or {}).get(area, "none"), 0):
            return area
    return None


def require(matrix: dict, area: str, level: str = "view") -> None:
    """Raise 403 unless the matrix grants at least `level` on `area`."""
    if RANK.get(matrix.get(area, "none"), 0) < RANK[level]:
        raise HTTPException(status_code=403, detail=f"You need {level} access to {area}.")
